File: Sources/FlowDictate/Services/OpenAIBackend.swift
import Foundation
import os.log

/// Remote transcription backend using OpenAI Audio Transcription API
/// Sends audio to OpenAI for transcription (requires API key)
final class OpenAIBackend: TranscriptionBackendProtocol {
    // MARK: - Constants

    private let apiURL = URL(string: "https://api.openai.com/v1/audio/transcriptions")!
    private let model = "whisper-1" // Can also use gpt-4o-transcribe when available

    // MARK: - Private State

    private let keychainService: KeychainService
    private let logger = Logger(subsystem: "com.flowdictate", category: "OpenAI")

    // MARK: - Initialization

    init(keychainService: KeychainService = KeychainService.shared) {
        self.keychainService = keychainService
    }

    // MARK: - TranscriptionBackendProtocol

    var isReady: Bool {
        get async {
            keychainService.getAPIKey() != nil
        }
    }

    func warmUp() async throws {
        // No warm-up needed for remote backend
        // Just verify API key exists
        guard keychainService.getAPIKey() != nil else {
            throw DictationError.apiKeyMissing
        }
    }

    func transcribe(audio: [Float], language: Language) async throws -> String {
        guard let apiKey = keychainService.getAPIKey() else {
            throw DictationError.apiKeyMissing
        }

        guard !audio.isEmpty else {
            throw DictationError.noAudioCaptured
        }

        logger.info("Starting remote transcription of \(audio.count) samples")
        let startTime = CFAbsoluteTimeGetCurrent()

        // Convert Float samples to WAV data
        let wavData = createWAVData(from: audio)

        // Create multipart form data request
        let boundary = UUID().uuidString
        var request = URLRequest(url: apiURL)
        request.httpMethod = "POST"
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")

        var body = Data()

        // Add model field
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"model\"\r\n\r\n".data(using: .utf8)!)
        body.append("\(model)\r\n".data(using: .utf8)!)

        // Add language field (if not auto)
        if let langCode = language.whisperCode {
            body.append("--\(boundary)\r\n".data(using: .utf8)!)
            body.append("Content-Disposition: form-data; name=\"language\"\r\n\r\n".data(using: .utf8)!)
            body.append("\(langCode)\r\n".data(using: .utf8)!)
        }

        // Add audio file
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"file\"; filename=\"audio.wav\"\r\n".data(using: .utf8)!)
        body.append("Content-Type: audio/wav\r\n\r\n".data(using: .utf8)!)
        body.append(wavData)
        body.append("\r\n".data(using: .utf8)!)

        // Close boundary
        body.append("--\(boundary)--\r\n".data(using: .utf8)!)

        request.httpBody = body

        // Send request
        do {
            let (data, response) = try await URLSession.shared.data(for: request)

            guard let httpResponse = response as? HTTPURLResponse else {
                throw DictationError.networkError("Invalid response")
            }

            guard httpResponse.statusCode == 200 else {
                let errorMessage = String(data: data, encoding: .utf8) ?? "Unknown error"
                logger.error("API error: \(httpResponse.statusCode) - \(errorMessage)")

                if httpResponse.statusCode == 401 {
                    throw DictationError.apiKeyMissing
                }
                throw DictationError.networkError("API error: \(httpResponse.statusCode)")
            }

            // Parse response
            let json = try JSONDecoder().decode(TranscriptionResponse.self, from: data)

            let elapsed = CFAbsoluteTimeGetCurrent() - startTime
            logger.info("Remote transcription completed in \(elapsed, format: .fixed(precision: 2))s")

            return json.text.trimmingCharacters(in: .whitespaces)
        } catch let error as DictationError {
            throw error
        } catch {
            logger.error("Network error: \(error.localizedDescription)")
            throw DictationError.networkError(error.localizedDescription)
        }
    }

    // MARK: - Private Methods

    /// Create WAV file data from Float samples at 16kHz mono
    private func createWAVData(from samples: [Float]) -> Data {
        let sampleRate: UInt32 = 16000
        let channels: UInt16 = 1
        let bitsPerSample: UInt16 = 16

        // Convert Float samples to Int16
        let int16Samples = samples.map { sample -> Int16 in
            let clamped = max(-1.0, min(1.0, sample))
            return Int16(clamped * Float(Int16.max))
        }

        let dataSize = UInt32(int16Samples.count * 2)
        let fileSize = dataSize + 36

        var header = Data()

        // RIFF header
        header.append("RIFF".data(using: .ascii)!)
        header.append(withUnsafeBytes(of: fileSize.littleEndian) { Data($0) })
        header.append("WAVE".data(using: .ascii)!)

        // fmt chunk
        header.append("fmt ".data(using: .ascii)!)
        header.append(withUnsafeBytes(of: UInt32(16).littleEndian) { Data($0) }) // chunk size
        header.append(withUnsafeBytes(of: UInt16(1).littleEndian) { Data($0) }) // audio format (PCM)
        header.append(withUnsafeBytes(of: channels.littleEndian) { Data($0) })
        header.append(withUnsafeBytes(of: sampleRate.littleEndian) { Data($0) })
        let byteRate = sampleRate * UInt32(channels) * UInt32(bitsPerSample / 8)
        header.append(withUnsafeBytes(of: byteRate.littleEndian) { Data($0) })
        let blockAlign = channels * (bitsPerSample / 8)
        header.append(withUnsafeBytes(of: blockAlign.littleEndian) { Data($0) })
        header.append(withUnsafeBytes(of: bitsPerSample.littleEndian) { Data($0) })

        // data chunk
        header.append("data".data(using: .ascii)!)
        header.append(withUnsafeBytes(of: dataSize.littleEndian) { Data($0) })

        // Audio data
        var audioData = Data(capacity: int16Samples.count * 2)
        for sample in int16Samples {
            withUnsafeBytes(of: sample.littleEndian) { audioData.append(contentsOf: $0) }
        }

        return header + audioData
    }
}

// MARK: - Response Types

private struct TranscriptionResponse: Codable {
    let text: String
}

File: flowdictate-tauri/src-tauri/src/transcription/openai_backend.rs
use reqwest::multipart;
use tracing::{error, info};

use crate::audio::wav::encode_wav;
use crate::credentials::KeyringService;
use crate::error::DictationError;
use crate::settings::Language;

use super::backend::TranscriptionBackend;

const API_URL: &str = "https://api.openai.com/v1/audio/transcriptions";
const MODEL: &str = "whisper-1";
const MAX_AUDIO_SIZE_BYTES: usize = 25 * 1024 * 1024;

pub struct OpenAIBackend {
    client: reqwest::Client,
    keyring: KeyringService,
}

impl OpenAIBackend {
    pub fn new(keyring: KeyringService) -> Self {
        let client = reqwest::Client::builder()
            .timeout(std::time::Duration::from_secs(60))
            .build()
            .expect("Failed to create HTTP client");

        Self { client, keyring }
    }
}

#[async_trait::async_trait]
impl TranscriptionBackend for OpenAIBackend {
    async fn is_ready(&self) -> bool {
        self.keyring.get_api_key().is_some()
    }

    async fn warm_up(&self) -> Result<(), DictationError> {
        if self.keyring.get_api_key().is_none() {
            return Err(DictationError::ApiKeyMissing);
        }
        Ok(())
    }

    async fn transcribe(&self, audio: &[f32], language: Language) -> Result<String, DictationError> {
        let api_key = self
            .keyring
            .get_api_key()
            .ok_or(DictationError::ApiKeyMissing)?;

        if audio.is_empty() {
            return Err(DictationError::NoAudioCaptured);
        }

        info!("Starting remote transcription of {} samples", audio.len());

        let wav_data = encode_wav(audio);

        if wav_data.len() > MAX_AUDIO_SIZE_BYTES {
            let size_mb = wav_data.len() as f64 / (1024.0 * 1024.0);
            return Err(DictationError::TranscriptionFailed(format!(
                "Audio file too large ({size_mb:.1}MB). Maximum is 25MB."
            )));
        }

        let file_part = multipart::Part::bytes(wav_data)
            .file_name("audio.wav")
            .mime_str("audio/wav")
            .unwrap();

        let mut form = multipart::Form::new()
            .text("model", MODEL)
            .part("file", file_part);

        if let Some(code) = language.whisper_code() {
            form = form.text("language", code.to_string());
        }

        let response = self
            .client
            .post(API_URL)
            .bearer_auth(&api_key)
            .multipart(form)
            .send()
            .await
            .map_err(|e| DictationError::NetworkError(e.to_string()))?;

        let status = response.status();
        if !status.is_success() {
            let body = response.text().await.unwrap_or_default();
            error!("OpenAI API error: {status} - {body}");
            if status.as_u16() == 401 {
                return Err(DictationError::ApiKeyMissing);
            }
            return Err(DictationError::NetworkError(format!("API error: {status}")));
        }

        #[derive(serde::Deserialize)]
        struct TranscriptionResponse {
            text: String,
        }

        let result: TranscriptionResponse = response
            .json()
            .await
            .map_err(|e| DictationError::TranscriptionFailed(format!("Failed to parse response: {e}")))?;

        info!("Remote transcription complete: {} chars", result.text.len());
        Ok(result.text.trim().to_string())
    }
}
